### ISSUE TITLE
Add interpolate method to Image class

### ISSUE THREAD

**Issue by Malcolmnixon** [godotengine/godot-proposals#3990]
### Describe the project you are working on

An extension to godot-xr-tools for drawing and erasing of markers on white-boards and windows

### Describe the problem or limitation you are having in your project

I'm able to use Image.blend_rect and/or Image.blend_rect_mask to "add" color to a white-board or window image texture. I can even use them to "clear" the white-board by painting white; however their underlying Color.blend operation makes it impossible to clear the window by blending transparency back into the image.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I would like to propose a new function Image.interpolate_rect(Image src, Image weight, Rect2 src_rect, Vector2 dst) which would apply the color transformation using:
    dst_color = dst_color.linear_interpolate(src_color, weight.alpha)

This would allow selective blending (controlled by the weight image alpha channel) that would allow blending in areas of a transparent source image onto the underlying image, rendering parts of it transparent again.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

The new proposed function would allow linearly interpolating a section of the existing image with a new image using the alpha channel of a weighting image.

The function would be very similar to Image::blend_rect_mask, but with the inner loop changed as follows:
```cpp
    for (int i = 0; i < dest_rect.size.y; i++) {
        for (int j = 0; j < dest_rect.size.x; j++) {
            int src_x = clipped_src_rect.position.x + j;
            int src_y = clipped_src_rect.position.y + i;

            // If the weights pixel is transparent then we skip it
            float wv = weight->get_pixel(src_x, src_y).a;
            if (wv != 0)
            {
                int dst_x = dest_rect.position.x + j;
                int dst_y = dest_rect.position.y + i;

                Color sc = img->get_pixel(src_x, src_y);
                Color dc = get_pixel(dst_x, dst_y);                
                dc = dc.linear_interpolate(sc, wv)
                set_pixel(dst_x, dst_y, dc;
            }
        }
    }
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I don't believe so - the performance of pixel-by-pixel image processing in scripts would be horrendous.

### Is there a reason why this should be core and not an add-on in the asset library?

This would be an additional method to an existing core class, and as such could not be located in a separate library.

**Comment by Calinou** [comment]
Note that all `linear_interpolate()` methods were renamed to `lerp()` in 4.0.

**Comment by Xrayez** [comment]
The [`ImageBlender`](https://goost.readthedocs.io/en/latest/classes/class_imageblender.html) in [Goost](https://goostengine.github.io/) has various OpenGL-derived blending functions, but I'm not sure whether you'd find them useful.

> An extension to godot-xr-tools for drawing and erasing of markers on white-boards and windows

For the use case of real-time drawing, I'd recommend using `Viewport` rendering functionality. You can set a viewport to never clear on next frame, and use `CanvasItem` blending materials to either draw or erase parts of the viewport. But this is of course more advanced approach.

**Comment by Malcolmnixon** [comment]
@Xrayez I switched my example project from Image to Viewport rendering, and used custom shaders to implement the desired blend/mix operations. It did require learning to do shaders, but it's something I've needed to do anyway and this is a case of using the right tool for the right job. The Viewport rendering leveraging the GPU is significantly faster than using the CPU and provides far more flexibility when compared to adding a library of custom blend/lerp functions to the Image class.

This alternate use of viewports for 2D GPU accelerated render surfaces does take a little getting used to, and perhaps warrants some improved documentation and examples, but that's always the case with documentation ;)

I withdraw my proposal as a technically superior solution already exists! Many thanks for the feedback everyone.
